sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,448 @@
+import org.apache.beam.sdk.schemas.logicaltypes.FixedBytes;
+import org.apache.beam.sdk.values.Row;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+/**
+ * Utilities to create {@link Iterable}s of Beam {@link Row} instances backed by Arrow record
+ * batches.
+ */
+@Experimental(Experimental.Kind.SCHEMAS)
+public class ArrowConversion {
+  /** Converts Arrow schema to Beam row schema. */
+  public static Schema toBeamSchema(org.apache.arrow.vector.types.pojo.Schema schema) {
+    return toBeamSchema(schema.getFields());
+  }
+
+  public static Schema toBeamSchema(List<org.apache.arrow.vector.types.pojo.Field> fields) {
+    Schema.Builder builder = Schema.builder();
+    for (org.apache.arrow.vector.types.pojo.Field field : fields) {
+      Field beamField = toBeamField(field);
+      builder.addField(beamField);
+    }
+    return builder.build();
+  }
+
+  /** Get Beam Field from Arrow Field. */
+  private static Field toBeamField(org.apache.arrow.vector.types.pojo.Field field) {
+    FieldType beamFieldType = toFieldType(field.getFieldType(), field.getChildren());
+    return Field.of(field.getName(), beamFieldType);
+  }
+
+  /** Converts Arrow FieldType to Beam FieldType. */
+  private static FieldType toFieldType(
+      org.apache.arrow.vector.types.pojo.FieldType arrowFieldType,
+      List<org.apache.arrow.vector.types.pojo.Field> childrenFields) {
+    FieldType fieldType =
+        arrowFieldType
+            .getType()
+            .accept(
+                new ArrowType.ArrowTypeVisitor<FieldType>() {
+                  @Override
+                  public FieldType visit(ArrowType.Null type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Struct type) {
+                    return FieldType.row(toBeamSchema(childrenFields));
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.List type) {
+                    checkArgument(
+                        childrenFields.size() == 1,
+                        "Encountered "
+                            + childrenFields.size()
+                            + " child fields for list type, expected 1");
+                    return FieldType.array(toBeamField(childrenFields.get(0)).getType());
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.FixedSizeList type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Union type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
```

Comment:
We now have OneOf - can you map to that?

Reply:
Ah yes I think that would be a good mapping.
Let's add a comment/jira to do this later though, we can verify BigQuery integration without it.